doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
```

Comment:
Do we really need both options (1) and (2)? When would someone choose one vs the other?

I agree with @pv that we should not recommend both (if possible), because it makes things even more confusing.

To give one concrete example, I don't think the implementation of ArrayLike [in my gist](https://gist.github.com/shoyer/36b84ab064f027df318c0b823558de24) would work with option (2), because that implementation of `__array_ufunc__` calls `getattr(ufunc, method)` instead of using `super()`, and hence can never return `NotImplemented`.

Even with that example aside, if both options can be made to work, we should prefer option (1) because it uses public API.